sampled_addmm: BSR support

nikitaved opened the pull request:
This PR implements a `sampled_addmm` kernel that works with a BSR mask.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101978
* #102095
* __->__ #101163

nikitaved committed:
sampled_addmm: BSR support

[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"




[ghstack-poisoned]

nikitaved commented:
@cpuhrsch , @amjames , this one is ready for review. I will update dispatch in the follow-up. Luckily, we do not have to deal with anything complicated.

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved commented on torch/sparse/_triton_ops.py:
```diff
@@ -88,11 +161,190 @@ def valid_grid_dim(g, mg):
+            else:
+                acc_block = alpha * acc_block + beta * tl.load(values_block_ptrs)
+
+            # write result
+            tl.store(values_block_ptrs, acc_block.to(values_ptr.dtype.element_ty))
+
+            # advance val/col_index ptrs to the next block in the row.
+            values_block_ptrs += values_nnz_stride
+            col_index_nnz_ptr += col_indices_stride
+
```
@cpuhrsch , @amjames , we can fuse this kernel with `softmax` and `dropout` all with the triton's high-level tools. I suggest we do it here in a follow-up, then I could work on backward.

nikitaved commented on test/test_sparse_csr.py:
```diff
@@ -3468,6 +3469,115 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
+                grid_size = (3,)
+                grid_gen = itertools.product(grid_size, repeat=2)
+                for grid in grid_gen:
+                    res_tri_grid = sampled_addmm(bsr, mat1, mat2, alpha=alpha, beta=beta)
```
Oh, right! Just missed it! Good catch, thanks!

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -3468,6 +3469,125 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
+        batches = [(), (2,), (2, 2)]
+        size = [128, 256, 0]
+
+        def sampled_addmm_ref(input, mat1, mat2, alpha, beta):
```
Could another way to implement this be `torch.addmm(input.to_dense(), mat1, mat2, alpha, beta).sparse_mask(input.to_sparse()).to_sparse_bsr(input.values().shape[-2:])`?

Sure it's less efficient, but a simpler reference implementation could give us more confidence?

nikitaved replied:
We test against CSR from `sparse.sampled_addmm` for more confidence. The proposed solution also only works for non-batched inputs, so, for simplicity, we still need to loop over batches...

cpuhrsch replied:
Yes, testing against CSR with float would work too. That's even simpler.

cpuhrsch reviewed (changes requested):
I'm a bit worried about the complexity of the reference addmm implementation just so we can have a simple comparison point.

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -3468,6 +3469,125 @@ def test_triton_bsr_dense_bmm_error_messages(self, device, dtype):
+                    res_csr = torch.sparse.sampled_addmm(csr, mat1csr, mat2csr, alpha=alpha, beta=beta)
+                    self.assertEqual(res_tri.to_dense(), res_csr.to_dense())
+
+                # Check grid consistency
```
What does this mean?

nikitaved replied:
I will update the comment to clarify on that.

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

cpuhrsch commented:
Looks like there's one lint error left otherwise this is good to go. Thanks for writing this!

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved committed:
Update on "sampled_addmm: BSR support"

This PR implements a `sampled_addmm` kernel that works with a BSR mask.




[ghstack-poisoned]

nikitaved commented:
@pytorchbot merge
